### DISCUSSION TITLE
SQL Alchemy for an online game?

### DISCUSSION THREAD

**Discussion by Lime-Parallelogram** [sqlalchemy/sqlalchemy#7343] [Usage Questions]
Hi there,
I would just like to ask for input on whether my usage of sqlalchemy within an online game is a valid one. I am attempting to use a mariadb database to keep track of all the play progression such as the amount of cash each player has etc. The game uses a flask webserver and the basic principle is that players sit on a webpage and after each round their new information is sent over via a socketio connection. At one point in the game, all of the players move from one webpage to another which seems to cause big problems for the SQLalchemy system - it gives me the below error:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/engineio/server.py", line 549, in _trigger_event
    return self.handlers[event](*args)
  File "/usr/local/lib/python3.9/site-packages/socketio/server.py", line 756, in _handle_eio_disconnect
    self._handle_disconnect(sid, '/')
  File "/usr/local/lib/python3.9/site-packages/socketio/server.py", line 663, in _handle_disconnect
    self._trigger_event('disconnect', '/', sid)
  File "/usr/local/lib/python3.9/site-packages/socketio/server.py", line 711, in _trigger_event
    return self.handlers[namespace][event](*args)
  File "/application/gameserver.py", line 68, in disconnect
    userLine = database.get_user(sid)
  File "/application/database.py", line 91, in decorator
    output = func(*args,**kwargs)
  File "/application/database.py", line 120, in get_user
    user = gameDB.query(activeUsers).filter(activeUsers.socketio_id==userID).first()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2810, in first
    return self.limit(1)._iter().first()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2894, in _iter
    result = self.session.execute(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1688, in execute
    conn = self._connection_for_bind(bind)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1529, in _connection_for_bind
    return self._transaction._connection_for_bind(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 721, in _connection_for_bind
    self._assert_active()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 594, in _assert_active
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: This session is in 'prepared' state; no further SQL can be emitted within this transaction.
```

This leads me to question whether this is the right tool for the job? Why do the requests not get stacked up and handled one after another? Is it possible that having two places trying to access the database is causing us problems as we have access from two separate docker containers.

Many thanks for you help;
This issue is proving quite a chore to work on.

**Comment by CaselIT** [comment]
Hi,

With the information you provided is impossible to say anything more that the session is being uses while prepared in two-phase commit.
The popular approach with web services is to use a single session for each request, closing the session at the end of the request.

If you can provide a reproducing example or at least more information we can try to offer support

**Reply by Lime-Parallelogram** [reply]
Many thanks for getting back to me.

The code we are working on is on GitHub here (https://github.com/A-Level-Personal-Code-Colab/PirateGame). I have read that using individual sessions for requests is a usual way to go for web services but the problem is we have two separate, independent modules both using the database, one for the webserver and one for the socketio server . These two things run independently in their own docker container but they both use all of the same database functions which are housed in this file (https://github.com/A-Level-Personal-Code-Collab/PirateGame/blob/main/app/database.py). Is this bad practise? I was looking to try and have a file that contained the models and have only this shared but I couldn't get this to work out for some reason.

**Reply by CaselIT** [reply]
Hi,

I've just looked quickly at database.py, and you have https://github.com/A-Level-Personal-Code-Collab/PirateGame/blob/main/app/database.py#L35 that creates a module level session.
You definitely don't want to do that when using multiple threads/greenlet. That seems where your problems come from.

I suggest you remove all module level session and instead use one session per request. if you use flask the flask-sqlalchemy plugin should manage that for you. for socketio it should be ok to have a session per websocket, but I don't have direct experience here.

**Reply by Lime-Parallelogram** [reply]
Ah, thanks. I will take a look at doing that and report on whether that fixed the issue. Are you suggesting that in, the socketio server, one should create a new session for each new event that comes in - I guess it would be no different really to each request in the web server. 

Thanks for the suggestion anyway!

**Reply by CaselIT** [reply]
> Are you suggesting that in, the socketio server, one should create a new session for each new event that comes in

Note that I've never used socketio, so just speaking on a websocket prospective, the usual use case is that a websocket is long lived where the client connects once then stays connected until it leaves the service; then while connected client and server share multiple messages.
Getting a new session and keeping it open for the duration of each websocket is probably not the optimal way since for each client you would need a connection to the db, and that's usually not something the db are really designed for when the client can be a large number. So probably creating a session for each message is the best solution, or as alternative creating a new session for each web socket but then after processing a message the server closes it, releasing the connection (this effectively the same thing as creating a new session for each message).

Since socketio uses websockets, I think you should be able to map the example above to its usage
